doc/man/man3/FuzzyCMeans.3
.TH "FuzzyCMeans" 3 "Tue Jul 13 2021" "Version v1.0" "Fuzzy C-means in native C++" \" -*- nroff -*-
.ad l
.nh
.SH NAME
FuzzyCMeans
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <FuzzyCMeans\&.h>\fP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBFuzzyCMeans\fP (std::string dataFilename, std::string paramsFilename)"
.br
.ti -1c
.RI "\fBFuzzyCMeans\fP (std::string dataFilename, int dimension_, int k_, int maxNumOfIterations_, int m_, double eps_)"
.br
.ti -1c
.RI "\fB~FuzzyCMeans\fP ()"
.br
.ti -1c
.RI "void \fBcomputeClusters\fP ()"
.br
.ti -1c
.RI "void \fBcomputeNewMembershipCoefficients\fP ()"
.br
.ti -1c
.RI "void \fBupdateMembershipCoefficients\fP ()"
.br
.ti -1c
.RI "bool \fBisOver\fP ()"
.br
.ti -1c
.RI "void \fBrunFuzzyCMeans\fP ()"
.br
.ti -1c
.RI "void \fBprintPartitionMatrix\fP ()"
.br
.ti -1c
.RI "void \fBprintCentroids\fP ()"
.br
.ti -1c
.RI "double \fBcomputeObjectiveFunction\fP ()"
.br
.in -1c
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "std::vector< \fBPoint\fP > \fBpoints\fP"
.br
.ti -1c
.RI "std::vector< \fBCluster\fP > \fBclusters\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
Definition at line 8 of file FuzzyCMeans\&.h\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "FuzzyCMeans::FuzzyCMeans (std::string dataFilename, std::string paramsFilename)"
Constructor 1\&. 
.PP
\fBParameters:\fP
.RS 4
\fIdataFilename\fP the name of the file in which the data to be clustered are stored\&. 
.br
\fIparamsFilename\fP the name of the configuration file\&. 
.RE
.PP

.PP
Definition at line 11 of file FuzzyCMeans\&.cpp\&.
.SS "FuzzyCMeans::FuzzyCMeans (std::string dataFilename, int dimension_, int k_, int maxNumOfIterations_, int m_, double eps_)"
Constructor 2\&. 
.PP
\fBParameters:\fP
.RS 4
\fIdataFilename\fP the name of the file in which the data to be clustered are stored\&. 
.br
\fIparamsFilename\fP the name of the configuration file\&. 
.RE
.PP

.PP
Definition at line 57 of file FuzzyCMeans\&.cpp\&.
.SS "FuzzyCMeans::~FuzzyCMeans ()"
Destructor\&. 
.PP
Definition at line 96 of file FuzzyCMeans\&.cpp\&.
.SH "Member Function Documentation"
.PP 
.SS "void FuzzyCMeans::computeClusters ()"
It computes the clusters in every iteration of the algorithm\&. 
.PP
Definition at line 108 of file FuzzyCMeans\&.cpp\&.
.SS "void FuzzyCMeans::computeNewMembershipCoefficients ()"
It computes the membership coefficients of points to clusters in every iteration of the algorithm\&. 
.PP
Definition at line 129 of file FuzzyCMeans\&.cpp\&.
.SS "double FuzzyCMeans::computeObjectiveFunction ()"
It computes the value of the objective function that controls the convergence of the algorithm\&. 
.PP
Definition at line 225 of file FuzzyCMeans\&.cpp\&.
.SS "bool FuzzyCMeans::isOver ()"
It checks if the algorithm converged\&. 
.PP
\fBReturns:\fP
.RS 4
true if the algorithm converged, false otherwise\&. 
.RE
.PP

.PP
Definition at line 166 of file FuzzyCMeans\&.cpp\&.
.SS "void FuzzyCMeans::printCentroids ()"
It prints the centroids to the standard output\&. 
.PP
Definition at line 215 of file FuzzyCMeans\&.cpp\&.
.SS "void FuzzyCMeans::printPartitionMatrix ()"
It prints the partition matrix to the standard output\&. 
.PP
Definition at line 205 of file FuzzyCMeans\&.cpp\&.
.SS "void FuzzyCMeans::runFuzzyCMeans ()"
It runs the Fuzzy c-means routine\&. 
.PP
Definition at line 180 of file FuzzyCMeans\&.cpp\&.
.SS "void FuzzyCMeans::updateMembershipCoefficients ()"
It updates the membership coefficients of points to clusters in every iteration of the algorithm\&. 
.PP
Definition at line 158 of file FuzzyCMeans\&.cpp\&.
.SH "Member Data Documentation"
.PP 
.SS "std::vector<\fBCluster\fP> FuzzyCMeans::clusters"
the clusters to be constructed 
.PP
Definition at line 24 of file FuzzyCMeans\&.h\&.
.SS "std::vector<\fBPoint\fP> FuzzyCMeans::points"
the data to be clustered 
.PP
Definition at line 22 of file FuzzyCMeans\&.h\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for Fuzzy C-means in native C++ from the source code\&.


FuzzyCMeans.cpp
#include "FuzzyCMeans.h"
#include "PropertiesParser.h"
#include "MathFunc.h"

#include <string>
#include <sstream>
#include <fstream>
#include <iostream>
#include <ctime>

FuzzyCMeans::FuzzyCMeans(std::string dataFilename, std::string paramsFilename)
{
	srand(static_cast<unsigned int>(time(NULL)));
	PropertiesParser pp(paramsFilename);
	dimension = pp.getPropertyAsIntOrDefaultTo("dimension", 4);
	k = pp.getPropertyAsIntOrDefaultTo("k", 200);
	maxNumOfIterations = pp.getPropertyAsIntOrDefaultTo("maxNumOfIterations", 1000);
	m = pp.getPropertyAsIntOrDefaultTo("m", 2);
	eps = pp.getPropertyAsDoubleOrDefaultTo("eps", 1e-6);
	
	std::string dataline;
	std::ifstream in;
	in.open(dataFilename);
	if (in.is_open())
	{
		while (std::getline(in, dataline))
		{
			std::stringstream ss(dataline);
			std::string item;
			StringVector items;
			while (getline(ss, item, ','))
			{
				items.push_back(item);
			}
			int ID = std::stoi(items[0]);
			Point point(ID, k);
			// TODO: instead of adding value in each iteration by calling addValue function,
			// initialize values vector in the constructor and then change the values in each iteration
			for (size_t i = 1; i < items.size(); ++i) 
			{
				point.addValue(std::stod(items[i]));
			}
			items.clear();
			points.push_back(point);
		}
		in.close();
	}
	for (size_t clusterID = 0; clusterID < k; ++clusterID)
	{
		DoubleVector initialCentroid;
		initialCentroid.assign(dimension, 0.0);
		Cluster cl(initialCentroid, clusterID);
		clusters.push_back(cl);
	}
}

FuzzyCMeans::FuzzyCMeans(std::string dataFilename, int dimension_, int k_, int maxNumOfIterations_, int m_, double eps_) :
	dimension(dimension_), k(k_), maxNumOfIterations(maxNumOfIterations_), m(m_), eps(eps_)
{
	srand(static_cast<unsigned int>(time(NULL)));
	
	std::string dataline;
	std::ifstream in;
	in.open(dataFilename);
	if (in.is_open())
	{
		while (std::getline(in, dataline))
		{
			std::stringstream ss(dataline);
			std::string item;
			StringVector items;
			while (getline(ss, item, ',')) 
			{
				items.push_back(item);
			}
			int ID = std::stoi(items[0]);
			Point point(ID, k);
			for (size_t i = 1; i < items.size(); ++i) 
			{
				point.addValue(std::stod(items[i]));
			}
			items.clear();
			points.push_back(point);
		}
		in.close();
	}
	for (size_t clusterID = 0; clusterID < k; ++clusterID)
	{
		DoubleVector initialCentroid;
		initialCentroid.assign(dimension, 0.0);
		Cluster cl(initialCentroid, clusterID);
		clusters.push_back(cl);
	}
}

FuzzyCMeans::~FuzzyCMeans()
{
	if (!points.empty())
	{
		points.clear();
	}
	if (!clusters.empty()) 
	{
		clusters.clear();
	}
}

void FuzzyCMeans::computeClusters()
{
	for (size_t clusterID = 0; clusterID < k; ++clusterID)
	{
		DoubleVector nominator;
		nominator.assign(dimension, 0.0);
		double denominator = 0.0;
		for (size_t pointID = 0; pointID < points.size(); ++pointID)
		{
			double wOldValue = points[pointID].getWOldValue(clusterID);
			wOldValue = std::pow(wOldValue, m);
			denominator += wOldValue;
			DoubleVector temp = mfnc::multiplyVectorByConstant(*(points[pointID].getValues()), wOldValue);
			mfnc::addToVector(nominator, temp);
			temp.clear();
		}
		DoubleVector centroid = mfnc::multiplyVectorByConstant(nominator, 1.0 / denominator);
		clusters[clusterID].setCentroid(centroid);
	}
}

void FuzzyCMeans::computeNewMembershipCoefficients()
{
	for (size_t pointID = 0; pointID < points.size(); ++pointID)
	{
		DoubleVector distances;
		for (size_t clusterID = 0; clusterID < clusters.size(); ++clusterID)
		{
			//dist = mfnc::ComputeEuclideanDistance(*(points[pointID].getValues()), *center);
			double dist = mfnc::computeEuclideanDistance(*(points[pointID].getValues()), *(clusters[clusterID].getCentroid()));
			distances.push_back(dist);
		}
		DoubleVector wNew;
		for (size_t firstOtherClusterID = 0; firstOtherClusterID < clusters.size(); ++firstOtherClusterID)
		{
			double currentClusterDistance = distances[firstOtherClusterID];
			double distancesSum = 0.0;
			for (size_t secondOtherClusterID = 0; secondOtherClusterID < clusters.size(); ++secondOtherClusterID)
			{
				double otherClusterDistance = distances[secondOtherClusterID];
				distancesSum += std::pow(currentClusterDistance / otherClusterDistance, 2 / (m - 1));
			}
			wNew.push_back(1.0 / distancesSum);
		}
		distances.clear();
		points[pointID].updateWNew(wNew);
		wNew.clear();
	}
}

void FuzzyCMeans::updateMembershipCoefficients()
{
	for (size_t pointID = 0; pointID < points.size(); ++pointID)
	{
		points[pointID].updateW();
	}
}

bool FuzzyCMeans::isOver()
{
	bool over = true;
	for (size_t pointID = 0; pointID < points.size(); ++pointID)
	{
		if (!points[pointID].compareW(eps))
		{
			over = false;
			break;
		}
	}
	return over;
}

void FuzzyCMeans::runFuzzyCMeans()
{
	int start = clock();
	size_t iter = 0;
	double objectiveValue = 0.0;
	for (iter = 0; iter < maxNumOfIterations; ++iter)
	{
		objectiveValue = computeObjectiveFunction();
		std::cout << "Iteration count = " << iter << ", obj. fcn = " << objectiveValue << std::endl;
		computeClusters();
		computeNewMembershipCoefficients();
		if (isOver())
		{
			break;
		}
		else
		{
			updateMembershipCoefficients();
		}
	}
	int stop = clock();
	double elapsedTime = (stop - start) / static_cast<double>(CLOCKS_PER_SEC);
	std::cout << "Fuzzy c-means finished after " << (iter + 1) << " iteration(s) (" << elapsedTime << " sec.).\n\n";
}

void FuzzyCMeans::printPartitionMatrix()
{
	std::cout << "Partition Matrix:\n\n";
	for (size_t pointID = 0; pointID < points.size(); ++pointID)
	{
		points[pointID].printMembershipCoefficients();
	}
	std::cout << std::endl;
}

void FuzzyCMeans::printCentroids()
{
	std::cout << "Centroids:\n\n";
	for (size_t clusterID = 0; clusterID < clusters.size(); ++clusterID)
	{
		clusters[clusterID].printCentroid();
	}
	std::cout << std::endl;
}

double FuzzyCMeans::computeObjectiveFunction()
{
	double objectiveValue = 0.0;
	for (size_t pointID = 0; pointID < points.size(); ++pointID)
	{
		for (size_t clusterID = 0; clusterID < clusters.size(); ++clusterID)
		{
			double dist = mfnc::computeEuclideanDistance(*(points[pointID].getValues()), *(clusters[clusterID].getCentroid()));
			dist = std::pow(dist, 2.0);
			objectiveValue += (std::pow(points[pointID].getWOldValue(clusterID), m) * dist);
		}
	}
	return objectiveValue;
}